src/Disks/ObjectStorages/MetadataStorageFromObject.cpp

```diff
@@ -0,0 +1,219 @@
+#include "MetadataStorageFromObject.h"
+#include <Disks/IDisk.h>
+#include <Disks/ObjectStorages/StaticDirectoryIterator.h>
+#include <Common/filesystemHelpers.h>
+#include <Common/logger_useful.h>
+#include <IO/WriteHelpers.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NOT_IMPLEMENTED;
+    extern const int LOGICAL_ERROR;
+}
+
+MetadataStorageFromObject::MetadataStorageFromObject(
+    ObjectStoragePtr object_storage_,
+    const std::string & object_storage_root_path_)
+    : object_storage(object_storage_)
+    , object_storage_root_path(object_storage_root_path_)
+{
+}
+
+MetadataTransactionPtr MetadataStorageFromObject::createTransaction() const
+{
+    return std::make_shared<MetadataStorageFromObjectTransaction>(*this);
+}
+
+const std::string & MetadataStorageFromObject::getPath() const
+{
+    return object_storage_root_path;
+}
+
+bool MetadataStorageFromObject::exists(const std::string & path) const
+{
+    auto object = StoredObject::create(*object_storage, fs::path(object_storage_root_path) / path);
+    return object_storage->exists(object);
+}
+
+bool MetadataStorageFromObject::isFile(const std::string &) const
+{
+    return true;
```

Comment:
It seems it's better to check here that the file exists.

Reply:
Okay, let's try (though it is not 100% and will produce excessive API calls).